crates/odrl/src/functions/json_serializer.rs

```diff
@@ -0,0 +1,601 @@
+    let mut prohibition_map = serde_json::Map::new();
+
+    let mut target_map = serde_json::Map::new();
+    if let Some(target_type) = &prohibition.target.edc_type {
+        target_map.insert("@type".to_string(), serde_json::json!(target_type));
+    }
+    if let Some(target_uid) = &prohibition.target.uid {
+        target_map.insert("uid".to_string(), serde_json::json!(target_uid));
+    }
+    if target_map.len() > 1 {
+        prohibition_map.insert("target".to_string(), serde_json::Value::Object(target_map));
+    } else {
+        prohibition_map.insert("target".to_string(), serde_json::json!(prohibition.target.uid.as_ref().unwrap_or(&String::new())));
+    }
+
+    if let Some(assigner) = &prohibition.assigner {
+        let assigner_map = serialize_party(assigner);
+        if assigner_map.len() > 1 {
+            prohibition_map.insert("assigner".to_string(), serde_json::Value::Object(assigner_map));
+        } else {
+            prohibition_map.insert("assigner".to_string(), serde_json::json!(assigner.uid.as_ref().unwrap_or(&String::new())));
+        }
+    }
+
+    if let Some(assignee) = &prohibition.assignee {
+        let assignee_map = serialize_party(assignee);
+        if assignee_map.len() > 1 {
+            prohibition_map.insert("assignee".to_string(), serde_json::Value::Object(assignee_map));
+        } else {
+            prohibition_map.insert("assignee".to_string(), serde_json::json!(assignee.uid.as_ref().unwrap_or(&String::new())));
+        }
+    }
+
+    if (prohibition.action.refinements.is_none()) && (prohibition.action.included_in.is_none()) && (prohibition.action.implies.len() == 0) {
+        prohibition_map.insert("action".to_string(), serde_json::json!(prohibition.action.name.clone()));
+    } else {
+        let action_map = serialize_action(&prohibition.action);
+        prohibition_map.insert("action".to_string(), serde_json::Value::Object(action_map));
+    }
+
+    if prohibition.constraints.len() != 0 {
+        let serialized_constraints = serialize_constraint(&prohibition.constraints);
+        prohibition_map.insert("constraint".to_string(), serde_json::json!(serialized_constraints));
+    }
+
+    if prohibition.remedies.len() != 0 {
+        let serialized_duties = serialize_duty(&prohibition.remedies);
+        prohibition_map.insert("remedy".to_string(), serde_json::json!(serialized_duties));
+    }
+
+    prohibition_map
+}
+
+
+fn serialize_obligation(obligation: &Obligation) -> serde_json::Map<String, serde_json::Value> {
+    let mut obligation_map = serde_json::Map::new();
+
+    let mut target_map = serde_json::Map::new();
+    if let Some(target_type) = &obligation.target.edc_type {
+        target_map.insert("@type".to_string(), serde_json::json!(target_type));
+    }
+    if let Some(target_uid) = &obligation.target.uid {
+        target_map.insert("uid".to_string(), serde_json::json!(target_uid));
+    }
+    if target_map.len() > 1 {
```

Comment:
This looks really repeating, wouldn't it be better if serialize_party would return a serde_json directly ?

Reply:
Do you mean assets (target)?
Parties are already processed via a serialize_party function. Nevertheless, I have now also outsourced the serialization of the assets to a function in order to reduce the redundancy (https://github.com/ArunaStorage/dspace-rs/pull/1/commits/783d13f5f4b69fe357e5c9ab6725d547dacd8fe7).